compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/ir/IrArrayBuilder.kt

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.jvm.ir
+
+import org.jetbrains.kotlin.backend.jvm.lower.inlineclasses.unboxInlineClass
+import org.jetbrains.kotlin.ir.builders.*
+import org.jetbrains.kotlin.ir.expressions.IrConstructorCall
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.expressions.IrFunctionAccessExpression
+import org.jetbrains.kotlin.ir.expressions.IrGetValue
+import org.jetbrains.kotlin.ir.types.IrType
+import org.jetbrains.kotlin.ir.types.classOrNull
+import org.jetbrains.kotlin.ir.util.constructors
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.ir.util.getPropertyGetter
+
+inline fun JvmIrBuilder.irArray(arrayType: IrType, block: IrArrayBuilder.() -> Unit): IrExpression =
+    IrArrayBuilder(this, arrayType).apply { block() }.build()
+
+fun JvmIrBuilder.irArrayOf(arrayType: IrType, elements: List<IrExpression> = listOf()): IrExpression =
+    irArray(arrayType) { elements.forEach { +it } }
+
+private class IrArrayElement(val expression: IrExpression, val isSpread: Boolean)
+
+class IrArrayBuilder(val builder: JvmIrBuilder, val arrayType: IrType) {
+    // We build unboxed arrays for inline classes (UIntArray, etc) by first building
+    // an unboxed array of the underlying primitive type, then coercing the result
+    // to the correct type.
+    val unwrappedArrayType = arrayType.unboxInlineClass()
```

Comment:
It's not clear should this utility perfrom any logic related to inline classes or it should be covered by inline class lowering.

Reply:
This is a special case for unsigned types that exists in the handling of varargs. Specifically, we can pass unsigned types as varargs through unwrapped arrays (e.g., `UIntArray` for `UInt` varargs). The array type in question is a wrapper around the underlying array of primitives (e.g., `IntArray`) and does not have a corresponding SpreadBuilder. In the example, we have to use `IntSpreadBuilder` directly and coerce the result to `UIntArray`.

We cannot implement this case in inline class lowering, since vararg lowering has to run early in the pipeline in order to implement the array copy optimizations (`testDoNotCopyImmediatelyCreatedArrays`) and inline class lowering does not actually change the IR types (inline class boxing and primitive type boxing are both implemented in the same way in `ExpressionCodegen`).